File: bun/src/cli/issue/history.ts
import type { Command } from "commander";
import { getClient } from "../../lib/client.ts";
import { printError, printPaginated, resolvePageSize } from "../../lib/output.ts";

export function registerHistory(issue: Command): void {
  const historyCmd = issue
    .command("history")
    .description("List activity history for an issue")
    .argument("<issue-id>", "Issue ID or key")
    .option("--limit <n>", "Limit results")
    .option("--cursor <token>", "Pagination cursor for next page");
  historyCmd.action(async (issueId: string) => {
    try {
      const client = getClient();
      const opts = historyCmd.opts<{ limit?: string; cursor?: string }>();
      const i = await client.issue(issueId);
      const history = await i.history({
        first: resolvePageSize(opts),
        after: opts.cursor,
      });
      const mapped = await Promise.all(
        history.nodes.map(async (h) => {
          const [
            actor,
            fromState,
            toState,
            fromAssignee,
            toAssignee,
            fromProject,
            toProject,
            addedLabels,
            removedLabels,
          ] = await Promise.all([
            h.actor,
            h.fromState,
            h.toState,
            h.fromAssignee,
            h.toAssignee,
            h.fromProject,
            h.toProject,
            h.addedLabels ?? Promise.resolve([]),
            h.removedLabels ?? Promise.resolve([]),
          ]);
          return {
            id: h.id,
            actor: actor ? { id: actor.id, name: actor.name } : null,
            fromState: fromState ? { id: fromState.id, name: fromState.name } : null,
            toState: toState ? { id: toState.id, name: toState.name } : null,
            fromAssignee: fromAssignee ? { id: fromAssignee.id, name: fromAssignee.name } : null,
            toAssignee: toAssignee ? { id: toAssignee.id, name: toAssignee.name } : null,
            fromPriority: h.fromPriority,
            toPriority: h.toPriority,
            fromEstimate: h.fromEstimate,
            toEstimate: h.toEstimate,
            fromTitle: h.fromTitle,
            toTitle: h.toTitle,
            fromDueDate: h.fromDueDate,
            toDueDate: h.toDueDate,
            fromProject: fromProject ? { id: fromProject.id, name: fromProject.name } : null,
            toProject: toProject ? { id: toProject.id, name: toProject.name } : null,
            addedLabels: addedLabels.map((l) => ({ id: l.id, name: l.name })),
            removedLabels: removedLabels.map((l) => ({ id: l.id, name: l.name })),
            updatedDescription: h.updatedDescription,
            archived: h.archived,
            trashed: h.trashed,
            autoArchived: h.autoArchived,
            autoClosed: h.autoClosed,
            createdAt: h.createdAt,
          };
        }),
      );
      printPaginated(mapped, history.pageInfo);
    } catch (err) {
      printError(err instanceof Error ? err.message : "List history failed");
    }
  });
}

File: internal/mappers/history.go
package mappers

import "github.com/shhac/lin/internal/linear"

// HistoryNode is the genqlient type for a single issue history entry.
type HistoryNode = linear.IssueHistoryIssueHistoryIssueHistoryConnectionNodesIssueHistory

// MapHistoryEntry converts a single issue history node into the standard output map.
func MapHistoryEntry(h HistoryNode) map[string]any {
	m := map[string]any{
		"id":                 h.Id,
		"createdAt":          h.CreatedAt,
		"fromPriority":       h.FromPriority,
		"toPriority":         h.ToPriority,
		"fromEstimate":       h.FromEstimate,
		"toEstimate":         h.ToEstimate,
		"fromTitle":          h.FromTitle,
		"toTitle":            h.ToTitle,
		"fromDueDate":        h.FromDueDate,
		"toDueDate":          h.ToDueDate,
		"updatedDescription": h.UpdatedDescription,
		"archived":           h.Archived,
		"trashed":            h.Trashed,
		"autoArchived":       h.AutoArchived,
		"autoClosed":         h.AutoClosed,
	}

	if h.Actor != nil {
		m["actor"] = map[string]any{"id": h.Actor.Id, "name": h.Actor.Name}
	}
	if h.FromState != nil {
		m["fromState"] = map[string]any{"id": h.FromState.Id, "name": h.FromState.Name}
	}
	if h.ToState != nil {
		m["toState"] = map[string]any{"id": h.ToState.Id, "name": h.ToState.Name}
	}
	if h.FromAssignee != nil {
		m["fromAssignee"] = map[string]any{"id": h.FromAssignee.Id, "name": h.FromAssignee.Name}
	}
	if h.ToAssignee != nil {
		m["toAssignee"] = map[string]any{"id": h.ToAssignee.Id, "name": h.ToAssignee.Name}
	}
	if h.FromProject != nil {
		m["fromProject"] = map[string]any{"id": h.FromProject.Id, "name": h.FromProject.Name}
	}
	if h.ToProject != nil {
		m["toProject"] = map[string]any{"id": h.ToProject.Id, "name": h.ToProject.Name}
	}

	addedLabels := make([]map[string]any, len(h.AddedLabels))
	for i, l := range h.AddedLabels {
		addedLabels[i] = map[string]any{"id": l.Id, "name": l.Name}
	}
	m["addedLabels"] = addedLabels

	removedLabels := make([]map[string]any, len(h.RemovedLabels))
	for i, l := range h.RemovedLabels {
		removedLabels[i] = map[string]any{"id": l.Id, "name": l.Name}
	}
	m["removedLabels"] = removedLabels

	return m
}
